torch/csrc/jit/python/python_sugared_value.cpp

```diff
@@ -217,6 +217,55 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
   return toSugaredValue(member, m, loc, /*is_constant=*/true);
 }
 
+ModuleValue::ModuleValue(
+    Value* self,
+    std::shared_ptr<ConcreteModuleType> concreteType,
```

Comment:
Why do need `hint` if we already have `concreteType` ?

Reply:
I consider these contained type hints to be properties of the parent module (after all, that's where the annotation lives), so they are in the concrete type for the parent, and not in the submodule concrete type. They need to be passed into the `ModuleValue` representing the submodule. 

Given that we only allow a module to appear once in an entire module hierarchy, I guess it's fine to associate a contained type hint with the submodule itself, and not with its parent.

Reply:
I agree that we only want to store it with the parent module where the annotation lives; I think we don't need to propagate it to the child module. Once we've indexed into the parent module, the output becomes annotated interface type

Reply:
I don't think I understand. If the contained type hint is in the parent's concrete type, and `parent.sub_module_dict` is desugared to a `ModuleValue` object containing the concrete type of `sub_module_dict` (which does not have the contained type hint), how do we figure out whether to allow non-static indexing in `getitem` for the `sub_module_dict`?

Reply:
Okay so you have the dict:
```
def forward(self, x, key: str)
    return self[key].forward(x)
```

When we index into `self` we know its concrete module type, and then we set the output of `self[key]` to be the contained interface type.

Reply:
What about the other unit test in this PR:
```
            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self.d[key].forward(x)
```
where `d` is a submodule that is a `ModuleDict` with a contained type hint annotated in the parent module?

Reply:
Yea, i think in this case `d` should have its contained Type Hint on its concrete module type

Reply:
But that concrete module type might be shared with other script modules instances to which the contained type hint does not apply.

Reply:
If the hint is part of the ConcreteModuleType itself, then it can't be shared with other module instances that it doesnt apply since we'll create a new type right ?

Reply:
```
    def get_or_create_concrete_type(self, nn_module):
        """
        Infer a ConcreteType from this `nn.Module` instance. Underlying JIT
        types are re-used if possible.
        """
        concrete_type_builder = infer_concrete_type_builder(nn_module)

        nn_module_type = type(nn_module)
        if nn_module_type not in self.type_store:
            self.type_store[nn_module_type] = []

        # Search the type store for an already-available JIT type
        known_types = self.type_store[nn_module_type]
        for known_type in known_types:
            if known_type.equals(concrete_type_builder):
                return known_type

        # We didn't find anything; generate a new JIT type from this concrete type
        concrete_type = concrete_type_builder.build()
        self.type_store[nn_module_type].append(concrete_type)
        return concrete_type
```

Ah, so you are saying that by setting the hint on some concrete type X, it will not be shared because 
```
        for known_type in known_types:
            if known_type.equals(concrete_type_builder):
                return known_type
```
will not find any type that matches?

Reply:
I think that makes sense.